### PR TITLE
HPCC 20752 Build and run HPCC in VS 2015

### PR DESCRIPTION
**Pull request by gaojx**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [x] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

File: dali/dafilesrv/dafilesrv.cpp
Comment (ghalliday) [inline dali/dafilesrv/dafilesrv.cpp]: more tabs

**Review by ghalliday — COMMENTED** [review]
A few comments from a brief initial scan.  I will look again in a bit.
Do we have details on how to build the dependent libraries in a wiki?

File: cmake_modules/FindLIBXML2.cmake
Comment (gaojx) [reply cmake_modules/FindLIBXML2.cmake]: Yes, will fix

File: common/remote/sockfile.cpp
Comment (gaojx) [reply common/remote/sockfile.cpp]: Agree.

File: configuration/configmgr/configmgrlib/EnvSupportLib.cpp
Comment (gaojx) [reply configuration/configmgr/configmgrlib/EnvSupportLib.cpp]: This is one issue with VS tab setting is not system wide: that is each time a new solution is generated, we needs to set the tab setting.

File: dali/dafilesrv/dafilesrv.cpp
Comment (gaojx) [reply dali/dafilesrv/dafilesrv.cpp]: fixed

File: cmake_modules/commonSetup.cmake
Comment (gaojx) [reply cmake_modules/commonSetup.cmake]: fixed

File: configuration/configmgr/configmgrlib/EnvSupportLib.cpp
Comment (ghalliday) [reply configuration/configmgr/configmgrlib/EnvSupportLib.cpp]: It doesn't align perfectly when I look at it in chrome.  Also, you can double check by trying to select a single character in the indentation.  If it is a tab it selects a block of characters instead.

Comment (ghalliday) [reply configuration/configmgr/configmgrlib/EnvSupportLib.cpp]: A pull request that replaces tabs in files that are predominately indented with spaces would be fine.  Quite a few of the tabs that I found with a similar search are in 3rd party code (e.g., cassandra/kafka) or in files submitted by external contributors where the whole file uses tabs.  I wouldn't replace those.

**Comment by ghalliday** [comment]
I tried building the branch on my machine, and came across the following problems:

a) The title of the commit message isn't quite the right format it should start HPCC-20752

b) Problems with references to the SSL ,library:
system\security\cryptohelper\cryptocommon.hpp(43): error C2065: 'EVP_PKEY_CTX': undeclared identifier
security\cryptohelper\cryptocommon.hpp(43): error C2065: 'EVP_PKEY_CTX_free': undeclared identifier

What is the mechanism for getting a compatible version of the SSL libraries.  (These will need to be installed by anyone using the system on windows, and the build box that creates the client tools.)

c) Various other errors.  I think these are all follow on errors because jlib cannot be built because of (b).

File: system/security/cryptohelper/pke.hpp
Comment (ghalliday) [inline system/security/cryptohelper/pke.hpp]: indent: tabs

**Review by ghalliday — COMMENTED** [review]
I think this is very close to being ready to take.  I think they are good changes, and will also make my life easier.
The main question is what to do about the external dependencies.  If we can come up with a list of packages, and how to download them/build them so that that platform can be built that would be great.

**Comment by ghalliday** [comment]
@gaojx a few more comments above.  I would also be tempted to target this to candidate-7.2.0, rather than master.  The earlier it is merged, the better.  It might possibly be a candidate for 7.0.2, but I will need to check it carefully and check with Richard before definitely deciding that.  I suspect the change in external dependencies will mean it should be 7.2.

File: cmake_modules/FindLIBXML2.cmake
Comment (gaojx) [reply cmake_modules/FindLIBXML2.cmake]: I don't have a good solution for this. Right now, we use the network share directory provided by Ming Wang (the build master), where all pre-compiled lib/dll are stored. Most of the libraries are already there. I made some change to it: add openssl 1.0.2h which build with current HPCC code, and add libxml2.

File: CMakeLists.txt
Comment (gaojx) [reply CMakeLists.txt]: Agree.

**Comment by gaojx** [comment]
This is address all the questions about getting the dependencies libraries on Windows.

Right now, Ming created a network shared directory in RELX internal network with all the prebuild lib/dll. 
You can get the URI and user credential to the shared directory. Just call the whole directory locally and point to in in you cmake command (-DEXTERNALS_DIRECTORY=...). 

The think the issue is more of legal question: if we are allowed to make it public (e.g., in a git repository), and/or distribute it as a package? I see websites where you can download prebuilt openssl libraries for various platform (include Windows x86 and x64).  I don't see why we can't do similarly legally? Maybe all we need is to have an approval from legal department,  add some legal disclaimer etc (whatever  the legal feels we need to do).

**Comment by ghalliday** [comment]
I am working on a solution for the dependencies using something called vcpkg.  So far it is looking promising, but requires a bit more work.

**Comment by ghalliday** [comment]
I am working on a solution for the dependencies using something called vcpkg.  So far it is looking promising, but requires a bit more work.

**Comment by gaojx** [comment]
I think we should merge the change now, so the other change #11910, #11926 can be merged.

**Comment by richardkchapman** [comment]
Should target 7.2.0, not earlier

**Comment by gaojx** [comment]
7.2.0 is fine. Is there anything I need to do to make the happens (i.e., set target to 7.2.0)?

**Comment by richardkchapman** [comment]
Edit the PR to set the target to be 7.2.0. You may also need to do

 rebase --onto candidate-7.2.0 master

and force-push your branch

**Comment by gaojx** [comment]
Rebased to candidate-7.2.0. Ready to merge.

File: CMakeLists.txt
Comment (richardkchapman) [inline CMakeLists.txt]: Why this change? Looks like indentation was correct before and is now wrong?

File: cmake_modules/commonSetup.cmake
Comment (richardkchapman) [inline cmake_modules/commonSetup.cmake]: The text of this option says the opposite of what the option does, I think. Seems to me the option's text should remain the same for both platforms, it's just the default value you are wanting to change.

File: cmake_modules/commonSetup.cmake
Comment (richardkchapman) [inline cmake_modules/commonSetup.cmake]: What about release builds?

Some of these warnings seem useful - what is the plan for re-enabling them?

File: common/remote/sockfile.cpp
Comment (richardkchapman) [inline common/remote/sockfile.cpp]: Any idea what changed here??

File: cmake_modules/commonSetup.cmake
Comment (richardkchapman) [inline cmake_modules/commonSetup.cmake]: typo outter should say outer

File: dali/dafilesrv/dafilesrv.cpp
Comment (richardkchapman) [inline dali/dafilesrv/dafilesrv.cpp]: ?

File: dali/dafilesrv/dafilesrv.cpp
Comment (richardkchapman) [inline dali/dafilesrv/dafilesrv.cpp]: this change seems somewhat unrelated

File: system/include/platform.h
Comment (richardkchapman) [inline system/include/platform.h]: This is not a good idea - should be calling FreeSharedObject (in jutil)

**Review by richardkchapman — CHANGES_REQUESTED** [review]
Mostly minor comments

**Comment by richardkchapman** [comment]
@gaojx

File: CMakeLists.txt
Comment (gaojx) [reply CMakeLists.txt]: It was a tab. Replaced with 4 spaces, should have been 8 spaces.

File: cmake_modules/commonSetup.cmake
Comment (gaojx) [reply cmake_modules/commonSetup.cmake]: The message tells where the 3rd party library is when the USE_NATIVE_LIBRARIES is set to default (which is OFF on Windows, and ON on other platform). I agree it is confusion, and will change it to be the same.

File: cmake_modules/commonSetup.cmake
Comment (gaojx) [reply cmake_modules/commonSetup.cmake]: Good point. Using CMAKE_CXX_FLAGS instead will work for all build types.
We can open one ticket per warning to re-enable them.

File: common/remote/sockfile.cpp
Comment (gaojx) [reply common/remote/sockfile.cpp]: The new version adds an EF BB BF header (means UTF8 encoding). It was probably added by some editor. Will remove it since no other cpp/hpp file has it.

File: cmake_modules/commonSetup.cmake
Comment (gaojx) [reply cmake_modules/commonSetup.cmake]: Fixed

File: dali/dafilesrv/dafilesrv.cpp
Comment (gaojx) [reply dali/dafilesrv/dafilesrv.cpp]: UTF8 BAM (byte order mark) header. Will remove.

File: dali/dafilesrv/dafilesrv.cpp
Comment (gaojx) [reply dali/dafilesrv/dafilesrv.cpp]: yes. will remove

File: system/include/platform.h
Comment (gaojx) [reply system/include/platform.h]: Good suggestion. Will do

File: dali/dafilesrv/dafilesrv.cpp
Comment (richardkchapman) [reply dali/dafilesrv/dafilesrv.cpp]: doesn't look like you removed it

File: cmake_modules/commonSetup.cmake
Comment (richardkchapman) [inline cmake_modules/commonSetup.cmake]: I think these changes are caused by BOM markers - probably best to avoid them

**Comment by richardkchapman** [comment]
A couple of minor comments. @ghalliday Are you happy with this, particularly regarding the external components part?

File: cmake_modules/commonSetup.cmake
Comment (gaojx) [reply cmake_modules/commonSetup.cmake]: Yes, this change remove the BOM markers.

File: dali/dafilesrv/dafilesrv.cpp
Comment (gaojx) [reply dali/dafilesrv/dafilesrv.cpp]: The variable is used by the code in the the source file, so we can't remove it.

**Comment by ghalliday** [comment]
@gaojx 
The external components need to be sensibly accessible.  I tried downloading them from the shared build directory, but it takes forever.

There seems to be a repository 
https://github.com/hpcc-systems/HPCC-Windows-Dependencies
which has some of the dependencies.  That should be updated with the new directory, so it is at least easy for anyone within the company to get the latest dependencies.  (Without this I will not be able to build the system any more.)

I'll review the rest in the meantime.

**Comment by ghalliday** [comment]
I think it is ready to merge once the dependencies are available in a repository.

**Pull request merged**
